test: Remove cilium DS before installing a new one

brb opened the pull request:
Previously, when installing Cilium via `ciliumInstallHelm()`, changes to ConfigMap were not applied if Cilium DaemonSet had already been installed.

An example of such occurrence in test logs:

    time="2020-02-01T21:01:58Z" level=debug msg="running command: kubectl apply --force=true -f cilium-15ef6310fc44caf0.yaml -n kube-system" cmd: "kubectl apply --force=true -f cilium-15ef6310fc44caf0.yaml -n
    kube-system" exitCode: 0 duration: 118.450977ms stdout:
    configmap/cilium-config configured
    serviceaccount/cilium unchanged
    serviceaccount/cilium-operator unchanged
    clusterrole.rbac.authorization.k8s.io/cilium unchanged
    clusterrole.rbac.authorization.k8s.io/cilium-operator unchanged
    clusterrolebinding.rbac.authorization.k8s.io/cilium unchanged
    clusterrolebinding.rbac.authorization.k8s.io/cilium-operator unchanged
    daemonset.apps/cilium unchanged
    deployment.apps/cilium-operator configured

Missing removal of the Cilium DS was causing some flakes, e.g. when the test suite `k8sT/external_ips.go` was ran before `k8sT/Services.go`. The
former suite in some test cases were using `--device=enp0s9` instead of `--device=enp0s8`, and `k8sT/Services.go` required the latter device setting (otherwise, a wrong IP addr for NodePort services was selected)

Fixes  #10031

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/10039)
<!-- Reviewable:end -->

brb committed:
test: Remove cilium DS before installing a new one

Previously, when installing Cilium via "ciliumInstallHelm()", changes to
ConfigMap were not applied if Cilium DaemonSet had already been
installed.

An example of such occurrence in test logs:

    time="2020-02-01T21:01:58Z" level=debug msg="running command: kubectl apply --force=true -f cilium-15ef6310fc44caf0.yaml -n kube-system" cmd: "kubectl apply --force=true -f cilium-15ef6310fc44caf0.yaml -n
    kube-system" exitCode: 0 duration: 118.450977ms stdout:
    configmap/cilium-config configured
    serviceaccount/cilium unchanged
    serviceaccount/cilium-operator unchanged
    clusterrole.rbac.authorization.k8s.io/cilium unchanged
    clusterrole.rbac.authorization.k8s.io/cilium-operator unchanged
    clusterrolebinding.rbac.authorization.k8s.io/cilium unchanged
    clusterrolebinding.rbac.authorization.k8s.io/cilium-operator unchanged
    daemonset.apps/cilium unchanged
    deployment.apps/cilium-operator configured

Missing removal of the Cilium DS were causing some flakes, e.g. when
the test suite k8sT/external_ips.go was ran before k8sT/Services.go. The
former in some test cases were using `--device=enp0s9` instead of
`--device=enp0s8`, and k8sT/Services.go required the latter device
setting.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

aanm commented on test/helpers/kubectl.go:
```diff
@@ -1396,6 +1396,11 @@ func (kub *Kubectl) ciliumInstallHelm(filename string, options map[string]string
+	// for the newly generated ConfigMap. Otherwise, the CM changes will stay
+	// inactive until each cilium-agent has been restarted.
+	_ = kub.DeleteResource("ds", fmt.Sprintf("-n %s cilium", GetCiliumNamespace(GetCurrentIntegration())))
+
```
All other places where we delete the cilium DS we are waiting for it to actually be deleted before continuing. In here we are not checking for it.

aanm reviewed (changes requested):
Also @raybejjani or @nebril PTAL

raybejjani commented on test/helpers/kubectl.go:
```diff
@@ -1396,6 +1396,14 @@ func (kub *Kubectl) ciliumInstallHelm(filename string, options map[string]string
+	// Remove cilium DS to ensure that new instances of cilium-agent are started
+	// for the newly generated ConfigMap. Otherwise, the CM changes will stay
+	// inactive until each cilium-agent has been restarted.
+	_ = kub.DeleteResource("ds", fmt.Sprintf("-n %s cilium", GetCiliumNamespace(GetCurrentIntegration())))
```
Why can't we just use `deleteCiliumDS`? Also, you should be able to use `CiliumNamespace` directly.

brb replied:
I've removed that function. Previously, it was defined in `test/k8sT/assertionHelpers.go`. So, using it from that module would introduce a circular dependency.

raybejjani commented on test/helpers/kubectl.go:
```diff
@@ -1404,6 +1412,35 @@ func (kub *Kubectl) ciliumInstallHelm(filename string, options map[string]string
+		default:
+		}
+
+		pods, err = kub.GetCiliumPodsContext(ctx, GetCiliumNamespace(GetCurrentIntegration()))
```
We should be able to use `CiliumNamespace` here

brb replied:
Why?

raybejjani commented on test/helpers/kubectl.go:
```diff
@@ -1404,6 +1412,35 @@ func (kub *Kubectl) ciliumInstallHelm(filename string, options map[string]string
+		if status == 0 {
+			return nil
+		}
+		time.Sleep(1 * time.Second)
```
We have `WithTimeout` helpers for this, I think

raybejjani commented on test/helpers/kubectl.go:
```diff
@@ -1404,6 +1412,35 @@ func (kub *Kubectl) ciliumInstallHelm(filename string, options map[string]string
 	return nil
 }
 
+func (kub *Kubectl) waitToDeleteCilium() error {
```
I haven't looked but it may be possible to approximate this with one of the WaitNPods type functions. Those may return when >=N is true, though, so that wouldn't work.

brb replied:
I moved this function from `test/k8sT/assertionHelpers.go`.

raybejjani reviewed (changes requested):
Could we just call the delete directly. This is the only cilium install function we have. If we're changing it then we need to make sure we aren't changing a test that assumes cilium won't be deleted first (like the upgrade test, but that may change soon as well). I would prefer a more explicit function call overall but I could be convinced that this one is ok.

brb committed:
test: Refactor deleteCiliumDS() usage

- Move the function to helpers/kubectl.go.
- Make sure that the function is called in AfterAll() of each suite.
- Remove redundant deleteCiliumDS() calls.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

raybejjani reviewed (approved):
Thank you!
